### PR TITLE
Menu improvements

### PR DESCRIPTION
**Pull request by simonhamp**
**Heads up, this introduces some significant breaking changes!**

Depends on https://github.com/NativePHP/electron/pull/139

We've been discussing internally for a while how the Menu API sucks a little. There are a few quirks, it doesn't really expose the stuff you probably want, and it just feels awkward, it doesn't feel very "Laravel".

1. You `new()` up a menu and then you have to `register()` it, which is what sets it as the application menu. It's not immediately obvious (imo) that you can omit `register()` and pass the resulting `Menu` to other things like Windows, the MenuBar and the Dock.

2. Defining items that should be in a list and can be nested inside each other as part of fluent chain feels really strange. You have to 'break' the chain (conceptually) in places that make it feel awkward and inconsistent, and the code ends up looking jarring rather than familiar.

3. The fluent chain is nice, but because we're immediately adding sparse instances to the chained `Menu` we lose access to the underlying `MenuItem` classes and so can't make deeper customisations easily.

4. It's currently impossible to change the application Menu once it's been defined.

So this is my attempt to improve on all of these issues by:

- Introducing a new `Menu` facade which wraps a new `MenuBuilder` class
    This menu builder let's you access the underlying `MenuItem` objects more easily and build your menus with a more expressive DX. More on this in a moment...
- More `MenuItem` types that map to standard OS functions (called `Roles`)
- Making `Link` menu items default to changing the currently-focused window's URL onclick without any extra wiring.
    If you want to open in the user's browser (the current default), you will be able to chain the new `->openInBrowser()` method to `Menu::link()`
- Making submenus (which are just `Menu` objects, which are a `MenuItem` too) definable in a more consistent, expressive way
- Adding a new `default()` helper to create the default menu even more easily
- Enhance event data coming back from menu item clicks, such as forwarding other data like combo keys
- Allow adding of an ID to `MenuItem` instances - useful in handling generic `MenuItemClicked` events on the Laravel side
- Allow the app menu to be completely recreated at any point in the app lifecycle 💪🏼

Here's what this looks like:

### Before

```php
use Native\Laravel\Menu\Menu;

Menu::new()
    ->appMenu()
    ->editMenu()
    ->register();
```

### After

```php
use Native\Laravel\Facades\Menu;

Menu::create(
    Menu::app(),
    Menu::edit(),
);
```

`Menu::create` builds the menu and registers it all in one shot.

### Before

```php
use Native\Laravel\Menu\Menu;

Menu::new()
    ->link('https://nativephp.com', 'Documentation');
```

### After

```php
use Native\Laravel\Facades\Menu;

Menu::make(
    Menu::link('https://nativephp.com', 'Documentation')
        ->openInBrowser(),
);
```

### Before
```php
use Native\Laravel\Menu\Menu;

Menu::new()
    ->appMenu()
    ->editMenu()
    ->viewMenu()
    ->windowMenu()
    ->register();
```

### After
```php
use Native\Laravel\Facades\Menu;

Menu::default();
```

`Menu::default()` allows you to easily return your menu to the default that would be provided if you didn't register a menu at all.

### Before
```php
use Native\Laravel\Menu\Menu;

Menu::new()
    ->appMenu()
    ->submenu(
        'My Submenu',
        Menu::new()
            ->link('https://nativephp.com', 'Documentation')
    )
    ->register();
```

### After
```php
use Native\Laravel\Facades\Menu;

Menu::create(
    Menu::app(),
    Menu::make(
        Menu::link('https://nativephp.com')
            ->label('Documentation')
            ->openInBrowser(),
        // or: Menu::link('https://nativephp.com', 'Documentation'),
    )->label('My Submenu'),
);
```

`Menu::make()` returns a `Native\Laravel\Menu\Menu` instance.

Many of the other helpers return a `MenuItem` subclass, from which you can chain calls to their other methods.

This gives you far more control over your menus. Building your menus has never been more straightforward.

### Before
```php
use Native\Laravel\Menu\Menu;

Menu::new()
    ->register();
```

### After
```php
use Native\Laravel\Facades\Menu;

Menu::create();
```

### CODE REVIEW

**Review by SRWieZ — CHANGES_REQUESTED** [review]
It's quite a change in syntax, but I believe this PR improves it.  
-  Some docblocks are incorrect or "incomplete"
-  I prefer using "hotkey" instead of "accelerator."  
- I got a bug with my event on a menu throwing an exception because of a missing named argument on my constructor. I think we should either ask devs to extend `MenuItemClicked` with their event or better create a contract that they should implement with the following constructor: `public function __construct(public array $item, public array $combo = [])`

Overall it's great!

Also we should add a note in the documentation stating that we cannot change the name of the first menu in macOS.  This tweet examples does not work, always returns the name of the app:
```php  
Menu::create(Menu::edit())  
```  
```php  
Menu::create(Menu::make()->label('My label'))  
```  

Capture d’écran 2024-11-22 à 20 41 33

**Comment by gwleuverink** [comment]
Honestly this works great! I've tinkered with it a bit now. 

One thing I might be missing. Context menu's were mentioned. Is it possible to trigger a menu on a button click? This doesn't work right now and could be a cool addition.

I've got a alpine magic I use for this right now. I'm happy to extract that to the Native helper if you want.

Right now it looks like this:

``` blade

<button
    type="button"
    x-on:click="
        $contextMenu([
            new MenuItem({
                label: 'Foo',
            }),
            new MenuItem({
                label: 'Bar',
            }),
        ])
    "
>

```

It does involve adding the MenuItem class to the window scope. But that can be namespaced under the Native helper if needed. 

I haven't tested this, but it might be feasible to add this to the php api instead. The code is pretty slim.

**Comment by gwleuverink** [comment]
Extracted to the Native helper I imagine something like this:

``` blade
<button
    type="button"
    onclick="Native.contextMenu([
        new MenuItem({
            label: 'Foo',
        }),
        new MenuItem({
            label: 'Bar',
        }),
    ])"
>
```

**Comment by gwleuverink** [comment]
> It does involve adding the MenuItem class to the window scope. But that can be namespaced under the Native helper if needed.

Turns out using `Menu.buildFromTemplate` I can construct this menu from a array of plain objects

**Comment by simonhamp** [comment]
@gwleuverink that context menu stuff looks super cool! Fancy wrapping whatever you've needed to modify there into a new PR?

**Comment by gwleuverink** [comment]
> @gwleuverink that context menu stuff looks super cool! Fancy wrapping whatever you've needed to modify there into a new PR?

Done! https://github.com/NativePHP/electron/pull/140

File: src/Facades/Menu.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+use Illuminate\Support\Facades\Facade;
+use Native\Laravel\Menu\Items\Checkbox;
+use Native\Laravel\Menu\Items\Label;
+use Native\Laravel\Menu\Items\Link;
+use Native\Laravel\Menu\Items\Radio;
+use Native\Laravel\Menu\Items\Role;
+use Native\Laravel\Menu\Items\Separator;
+
+/**
+ * @method static \Native\Laravel\Menu\Menu make(MenuItem ...$items)
```
[the remark this answers is not in the document]
Comment (simonhamp) [reply src/Facades/Menu.php:14]:
```suggestion
 * @method static \Native\Laravel\Menu\Menu make(\Native\Laravel\Menu\Items\MenuItem ...$items)
```

File: src/Facades/Menu.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * @method static Checkbox checkbox(string $label, bool $checked = false, ?string $hotkey = null)
+ * @method static Label label(string $label)
+ * @method static Link link(string $url, string $label = null, ?string $hotkey = null)
```
[the remark this answers is not in the document]
Comment (simonhamp) [reply src/Facades/Menu.php:17]: Ah, I think you mean in the `MenuItem` class, but the Facade instantiates the `MenuBuilder` - there it is called `$hotkey`.

I'm not going to change the `MenuItem` class, as that maps more closely to the underlying Electron implementation and feels clearer as `$accelerator` to me in that context.

I prefer to create an alias, per your next point

File: src/Events/Menu/MenuItemClicked.php [hunk 3]
```diff
@@ -12,7 +12,7 @@ class MenuItemClicked implements ShouldBroadcastNow
 {
     use Dispatchable, InteractsWithSockets, SerializesModels;
 
-    public function __construct(public array $item) {}
+    public function __construct(public array $item, public array $combo = []) {}
```
[the remark this answers is not in the document]
Comment (simonhamp) [reply src/Events/Menu/MenuItemClicked.php:15]: This comes straight from Electron and I don't want the docblock structure to be wrong if/when they change things, as that would be more confusing imo. Perhaps at some point we create our own abstraction around this? 🤷🏻‍♂️

**Pull request merged**